File: backend/app/routers/export_mappings.py
# backend/app/routers/export_mappings.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.auth import get_current_active_user
from app.models import User
from app.repositories.export_mapping_repository import ExportMappingRepository
from app.schemas.export_mapping import (
    ExportMappingCreate,
    ExportMappingResponse,
    ExportMappingUpdate,
)
from app.services.export_mapping_service import ExportMappingService
from app.services.exceptions import ConflictError, NotFoundError

router = APIRouter(
    prefix="/projects/{project_id}/export-mappings",
    tags=["export-mappings"],
)


def get_service(db: AsyncSession = Depends(get_db)) -> ExportMappingService:
    return ExportMappingService(ExportMappingRepository(db))


@router.get("", response_model=list[ExportMappingResponse])
async def list_export_mappings(
    project_id: UUID,
    data_store_id: UUID = Query(...),
    current_user: User = Depends(get_current_active_user),
    service: ExportMappingService = Depends(get_service),
):
    return await service.list_by_store(project_id, data_store_id)


@router.post("", response_model=ExportMappingResponse, status_code=status.HTTP_201_CREATED)
async def create_export_mapping(
    project_id: UUID,
    data: ExportMappingCreate,
    current_user: User = Depends(get_current_active_user),
    service: ExportMappingService = Depends(get_service),
):
    try:
        return await service.create(project_id, data)
    except ConflictError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.put("/{mapping_id}", response_model=ExportMappingResponse)
async def update_export_mapping(
    project_id: UUID,
    mapping_id: UUID,
    data: ExportMappingUpdate,
    current_user: User = Depends(get_current_active_user),
    service: ExportMappingService = Depends(get_service),
):
    try:
        return await service.update(mapping_id, project_id, data)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.delete("/{mapping_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_export_mapping(
    project_id: UUID,
    mapping_id: UUID,
    current_user: User = Depends(get_current_active_user),
    service: ExportMappingService = Depends(get_service),
):
    try:
        await service.delete(mapping_id, project_id)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))

File: frontend/src/api/exportMappings.ts
// frontend/src/api/exportMappings.ts
import apiClient from './client'
import type { ExportMapping, ExportMappingCreate, ExportMappingUpdate } from '@/types/exportMapping'

function toSnakeCase(data: ExportMappingCreate | ExportMappingUpdate): Record<string, unknown> {
  const result: Record<string, unknown> = {}
  if ('dataStoreId' in data && data.dataStoreId !== undefined) {
    result.data_store_id = data.dataStoreId
  }
  if (data.name !== undefined) result.name = data.name
  if (data.fieldMapping !== undefined) result.field_mapping = data.fieldMapping
  return result
}

function fromApi(raw: Record<string, unknown>): ExportMapping {
  return {
    id: raw.id as string,
    projectId: raw.project_id as string,
    dataStoreId: raw.data_store_id as string,
    name: raw.name as string,
    fieldMapping: raw.field_mapping as ExportMapping['fieldMapping'],
    createdAt: raw.created_at as string,
    updatedAt: raw.updated_at as string,
  }
}

export async function listExportMappings(
  projectId: string,
  dataStoreId: string
): Promise<ExportMapping[]> {
  const response = await apiClient.get<Record<string, unknown>[]>(
    `/projects/${projectId}/export-mappings`,
    { params: { data_store_id: dataStoreId } }
  )
  return response.data.map(fromApi)
}

export async function createExportMapping(
  projectId: string,
  data: ExportMappingCreate
): Promise<ExportMapping> {
  const response = await apiClient.post<Record<string, unknown>>(
    `/projects/${projectId}/export-mappings`,
    toSnakeCase(data)
  )
  return fromApi(response.data)
}

export async function updateExportMapping(
  projectId: string,
  mappingId: string,
  data: ExportMappingUpdate
): Promise<ExportMapping> {
  const response = await apiClient.put<Record<string, unknown>>(
    `/projects/${projectId}/export-mappings/${mappingId}`,
    toSnakeCase(data)
  )
  return fromApi(response.data)
}

export async function deleteExportMapping(
  projectId: string,
  mappingId: string
): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/export-mappings/${mappingId}`)
}
